### PR TITLE
Use all UTxOs when getting collateral from CIP-30 wallets

### LINKED ISSUE #723: Fix our use of CIP-30 getWalletCollateral
According to https://cips.cardano.org/cips/cip30/#apigetcollateralparamsamountcborcoinpromisetransactionunspentoutputnull the getCollateral call should return a number of utxos

>  that are required to reach AT LEAST the combined ADA value target specified in amount

But we are [only getting the first UTxO, ignoring the rest](https://github.com/Plutonomicon/cardano-transaction-lib/blob/bc7dca25ab450594e2bacc36a505ec17fa6939f3/src/Wallet/Cip30.js#L8). Hence the undercollateralized tx error (`collateralTooSmall`) that looks like this and is triggered in some conditions:

```
{"type":"jsonwsp/response","version":"1.0","servicename":"ogmios","methodname":"SubmitTx","result":{"SubmitFail":[{"collateralTooSmall":{"requiredCollateral":2249043,"actualCollateral":2146325}}]},"reflection":"SubmitTx-l5h435fz"}
```

### PR DESCRIPTION
**Pull request by klntsky**
Closes #723

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [ ] **All** existing examples have been run locally against a fully-synced testnet node
- [x] The integration and unit tests have been run (`npm run test`) locally
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`)

### CODE REVIEW

File: src/Wallet/Cip30.purs [hunk 1]
```diff
@@ -82,14 +81,15 @@ getWalletAddress :: Cip30Connection -> Aff (Maybe Address)
 getWalletAddress conn = fromHexString _getAddress conn <#>
   (_ >>= addressFromBytes <<< rawBytesAsCborBytes)
 
-getCollateral :: Cip30Connection -> Aff (Maybe TransactionUnspentOutput)
-getCollateral conn = fromMaybeHexString getCip30Collateral conn >>=
+getCollateral :: Cip30Connection -> Aff (Maybe (Array TransactionUnspentOutput))
+getCollateral conn = fromMaybeHexStrings getCip30Collateral conn >>=
   case _ of
     Nothing -> pure Nothing
```
Comment (jy14898) [inline src/Wallet/Cip30.purs:87]:
```suggestion
  (traverse >>> traverse)
```

File: src/Wallet/Cip30.purs [hunk 2]
```diff
@@ -82,14 +81,15 @@ getWalletAddress :: Cip30Connection -> Aff (Maybe Address)
...
-    Just bytes -> do
-      liftEffect $
-        Deserialization.UnspentOuput.convertUnspentOutput
-          <$> fromBytesEffect (unwrap bytes)
+    Just collateralUtxos -> liftEffect do
+      Just <$> for collateralUtxos \bytes -> do
+        maybe (throw "Unable to convert UTxO") pure =<<
+          Deserialization.UnspentOuput.convertUnspentOutput
+            <$> fromBytesEffect (unwrap bytes)
```
Comment (jy14898) [inline src/Wallet/Cip30.purs:92]:
```suggestion
    \bytes -> do
      utxo <- liftEffect $ fromBytesEffect $ unwrap bytes
      let utxo' = Deserialization.UnspentOuput.convertUnspentOutput utxo
      liftMaybe (error "Unable to convert UTxO") utxo'
```

File: src/Wallet/Cip30.purs [hunk 3]
```diff
@@ -119,12 +119,12 @@ fromHexString
...
-  :: (Cip30Connection -> Effect (Promise (Maybe String)))
+fromMaybeHexStrings
+  :: (Cip30Connection -> Effect (Promise (Maybe (Array String))))
   -> Cip30Connection
-  -> Aff (Maybe RawBytes)
-fromMaybeHexString act =
-  map (flip bind hexToRawBytes) <<< Promise.toAffE <<< act
+  -> Aff (Maybe (Array RawBytes))
+fromMaybeHexStrings act =
+  map (join <<< map (traverse hexToRawBytes)) <<< Promise.toAffE <<< act
```
Comment (jy14898) [inline src/Wallet/Cip30.purs:127]:
```suggestion
fromMaybeHexStrings mhs =
  Promise.toAffE mhs <#> (_ >>= traverse hexToRawBytes)
```

**Review by jy14898 — CHANGES_REQUESTED** [review]
Generally looks good to me, I've requested some changes to improve the readability (imo). It might be worth doing a similar thing for `signTx`/`fromHexString` here or in the future

File: src/Wallet/Cip30.purs (hunk 2 above)
Comment (ngua) [reply src/Wallet/Cip30.purs:92]: I don't personally find this more readable

Comment (jy14898) [reply src/Wallet/Cip30.purs:92]: Fair enough

Comment (jy14898) [reply src/Wallet/Cip30.purs:92]: At minimum I think `fromMaybeHexString` is doing more than it needs to, `<<< act` could be completely removed and the application of Cip30Connection moved to the call site

Comment (klntsky) [reply src/Wallet/Cip30.purs]: I refactored it, but another way.

File: CHANGELOG.md [hunk 4]
```diff
@@ -16,8 +16,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
   - [Fixed](#fixed)
 - [[1.1.0] - 2022-06-30](#110---2022-06-30)
   - [Fixed](#fixed-1)
-- [[1.0.1] - 2022-06-17](#101---2022-06-17)
+- [[1.1.0] - 2022-06-30](#110---2022-06-30)
```
Comment (jy14898) [inline CHANGELOG.md:19]: Something odd has happened here, we have two 1.1.0 entries?

File: src/Wallet/Cip30.purs [hunk 5]
```diff
@@ -82,12 +81,15 @@ getWalletAddress :: Cip30Connection -> Aff (Maybe Address)
 getWalletAddress conn = fromHexString _getAddress conn <#>
   (_ >>= addressFromBytes <<< rawBytesAsCborBytes)
 
-getCollateral :: Cip30Connection -> Aff (Maybe TransactionUnspentOutput)
-getCollateral conn = fromMaybeHexString getCip30Collateral conn >>=
-  case _ of
-    Nothing -> pure Nothing
-    Just bytes -> do
-      liftEffect $
+getCollateral :: Cip30Connection -> Aff (Maybe (Array TransactionUnspentOutput))
```
Comment (adamczykm) [inline src/Wallet/Cip30.purs:84]: A commentary could be useful here. Also when does it return `Nothing` and when does it throw

File: src/Wallet/Cip30.purs [hunk 6]
```diff
@@ -82,12 +81,15 @@ getWalletAddress :: Cip30Connection -> Aff (Maybe Address)
...
+getCollateral conn = do
+  mbUtxoStrs <- toAffE $ getCip30Collateral conn
+  let
+    (mbUtxoBytes :: Maybe (Array RawBytes)) =
+      join $ map (traverse hexToRawBytes) mbUtxoStrs
+  liftEffect $ for mbUtxoBytes \collateralUtxos -> do
```
Comment (adamczykm) [inline src/Wallet/Cip30.purs:90]: a short comment here to sum up the next 4 lines would be nice

**Review by adamczykm — CHANGES_REQUESTED** [review]
Looks nice, but the code could use some commentary and docstrings to reduce readers' cognitive load.

**Review by adamczykm — APPROVED** [review]
thanks

**Review by jy14898 — APPROVED** [review]

**Pull request merged**